File: packages/core-py/src/ebb_ai/sign.py
"""Ed25519 receipt signing (v0.11+, opt-in).

Python mirror of ``packages/core-ts/src/sign.ts``. Where the TS package
uses Node's built-in ``crypto`` (no deps), the Python build pulls in
``cryptography`` only when the caller opts in via the ``signing`` extras
package — keeping ``pip install ebb-ai`` lightweight for users who only
want the scheduler.

Install with signing enabled:

    pip install "ebb-ai[signing]"

Without that extra, importing this module is fine, but :func:`sign_receipt`
raises :class:`SigningNotInstalled`. Verification (:func:`verify_receipt`)
of pre-signed receipts also requires the extra. The scheduler itself
silently falls back to writing unsigned receipts when the extra is
absent, so the default install path remains v0.10-compatible.

Cross-language canonical form (audit §0.3): there is exactly ONE
canonical signing payload shared by both ports — the **camelCase wire
rendering** of the receipt (the TS shape; ``ebb verify`` is the
consumer). :func:`sign_receipt` converts the snake_case receipt dict to
camelCase algorithmically before canonicalization, and numbers are
serialized ECMAScript-style (RFC 8785 / JCS: ``2.0`` → ``"2"``,
``1e-07`` → ``"1e-7"``) via :func:`es_number` so the canonical bytes
match ``JSON.stringify`` exactly. The RETURNED receipt dict still
carries snake_case keys and snake_case signature fields (``signature``,
``signer_public_key``, ``signed_at``) — Python storage and consumers
are unchanged.

``signed_at`` is computed BEFORE canonicalization and is covered by the
signature (v0.12+); only ``signature`` and ``signer_public_key`` sit
outside the signed payload. Replay defence still requires ledger-side
uniqueness (e.g. unique ``task_id``): a validly signed receipt can be
presented twice.

:func:`verify_receipt` accepts a receipt in EITHER key rendering
(camelCase TS receipts included) and falls back to the legacy v0.11
canonical forms for receipts signed by older releases.

The shared fixture
``packages/core-ts/test/fixtures/cross-lang-receipt-vectors.json`` pins
the canonical bytes; both language test suites assert against it
byte-for-byte.

See ``packages/core-ts/src/sign.ts`` for the full design rationale and
key lifecycle.
"""

from __future__ import annotations

import base64
import contextlib
import json
import math
import os
import re
import stat
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path
from typing import Any, Literal

try:
    from cryptography.hazmat.primitives import serialization
    from cryptography.hazmat.primitives.asymmetric import ed25519

    _HAS_CRYPTO = True
except ImportError:  # pragma: no cover — exercised in non-extra installs
    _HAS_CRYPTO = False


class SigningNotInstalled(ImportError):
    """Raised when signing is invoked without the ``signing`` extras installed."""

    def __init__(self) -> None:
        super().__init__(
            "Receipt signing requires the `cryptography` library. "
            "Install via `pip install 'ebb-ai[signing]'`."
        )


def is_signing_available() -> bool:
    """Return True iff the ``cryptography`` extra is installed."""
    return _HAS_CRYPTO


VerifyOutcome = Literal["valid", "tampered", "legacy-unsigned", "key-mismatch"]


@dataclass(frozen=True, slots=True)
class SigningKeyPair:
    """An Ed25519 keypair plus the filesystem path the private key lives at.

    ``public_key_base64`` is the raw 32-byte public key — same encoding
    used by the TS port — suitable for embedding directly on a receipt.
    """

    private_key: Any
    public_key: Any
    public_key_base64: str
    private_key_path: str


@dataclass(frozen=True, slots=True)
class VerifyResult:
    outcome: VerifyOutcome
    reason: str
    signer_public_key: str | None = None


def default_signing_key_path() -> Path:
    """Default ``~/.ebb-ai/signing.key`` location."""
    return Path.home() / ".ebb-ai" / "signing.key"


def _write_bytes_atomic(path: Path, data: bytes, mode: int | None = None) -> None:
    """Write ``data`` atomically: same-directory temp file + ``os.replace``.

    Readers never observe a partial file; concurrent first-callers
    converge on one winner. ``mode`` (if given) is applied to the temp
    file BEFORE the rename so the final file never exists with loose
    permissions.
    """
    tmp = path.with_name(f"{path.name}.tmp-{os.getpid()}")
    try:
        tmp.write_bytes(data)
        if mode is not None:
            # Best-effort chmod — some filesystems (WSL/Windows mounts) refuse it.
            with contextlib.suppress(OSError):
                os.chmod(tmp, mode)
        os.replace(tmp, path)
    except BaseException:
        with contextlib.suppress(OSError):
            tmp.unlink()
        raise


def load_or_create_signing_key(
    private_key_path: str | os.PathLike[str] | None = None,
) -> SigningKeyPair:
    """Load the local signing keypair, generating + persisting one on first call.

    Mirrors the TS ``loadOrCreateSigningKey()``: idempotent and safe under
    concurrent first-call (files are written atomically via
    write-temp-then-rename; the winning writer's key is what both
    callers converge on next load).
    """
    if not _HAS_CRYPTO:
        raise SigningNotInstalled()
    path = Path(private_key_path) if private_key_path else default_signing_key_path()
    public_path = path.with_suffix(path.suffix + ".pub")

    if path.exists():
        pem = path.read_bytes()
        private_key = serialization.load_pem_private_key(pem, password=None)
        public_key = private_key.public_key()
    else:
        private_key = ed25519.Ed25519PrivateKey.generate()
        public_key = private_key.public_key()
        private_pem = private_key.private_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PrivateFormat.PKCS8,
            encryption_algorithm=serialization.NoEncryption(),
        )
        public_pem = public_key.public_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PublicFormat.SubjectPublicKeyInfo,
        )
        path.parent.mkdir(parents=True, exist_ok=True)
        _write_bytes_atomic(path, private_pem, mode=stat.S_IRUSR | stat.S_IWUSR)
        _write_bytes_atomic(public_path, public_pem)

    raw_pub = public_key.public_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PublicFormat.Raw,
    )
    return SigningKeyPair(
        private_key=private_key,
        public_key=public_key,
        public_key_base64=base64.b64encode(raw_pub).decode("ascii"),
        private_key_path=str(path),
    )


# Fields excluded from the signed payload — they are the proof itself.
# NOTE: ``signedAt`` is NOT excluded: since v0.12 it is covered by the
# signature (audit §0.8). These are the *normalized* (camelCase) names.
_SIGNATURE_FIELDS = frozenset({"signature", "signerPublicKey"})

# Every signature-machinery key, in both wire renderings.
_SIGNATURE_FIELDS_ANY_RENDERING = frozenset(
    {"signature", "signerPublicKey", "signer_public_key", "signedAt", "signed_at"}
)

_ES_SAFE_INT = 2**53


def es_number(value: int | float) -> str:
    """Serialize a number exactly like ECMAScript ``JSON.stringify``.

    RFC 8785 (JCS) number formatting: shortest round-trip digits,
    integers without a trailing ``.0`` (``2.0`` → ``"2"``), positional
    notation for magnitudes in ``[1e-6, 1e21)``, ES exponent style
    outside it (``1e-07`` → ``"1e-7"``, ``1e21`` → ``"1e+21"``), and
    ``-0.0`` → ``"0"``. Ints beyond 2**53 are coerced through float —
    JavaScript would have parsed them as IEEE-754 doubles anyway.
    """
    if isinstance(value, bool):  # bool is an int subclass; not a number here.
        raise TypeError("es_number: booleans are not numbers")
    if isinstance(value, int):
        if -_ES_SAFE_INT <= value <= _ES_SAFE_INT:
            return str(value)
        value = float(value)  # beyond safe range JS sees a double
    if not math.isfinite(value):
        raise ValueError(f"canonicalize: refusing non-finite number {value}")
    if value == 0:
        return "0"  # covers -0.0 → "0", matching JSON.stringify(-0)

    # repr() gives the shortest round-trip decimal form; re-render it
    # with ECMAScript's Number::toString notation rules (ECMA-262
    # §6.1.6.1.20). Represent value as 0.<digits> * 10**n.
    text = repr(value)
    sign = ""
    if text.startswith("-"):
        sign, text = "-", text[1:]
    mantissa, _, exp_text = text.partition("e")
    exponent = int(exp_text) if exp_text else 0
    int_part, _, frac_part = mantissa.partition(".")
    combined = int_part + frac_part
    digits = combined.lstrip("0")
    leading_zeros = len(combined) - len(digits)
    digits = digits.rstrip("0")
    n = len(int_part) - leading_zeros + exponent
    k = len(digits)

    if k <= n <= 21:
        body = digits + "0" * (n - k)
    elif 0 < n <= 21:
        body = f"{digits[:n]}.{digits[n:]}"
    elif -6 < n <= 0:
        body = "0." + "0" * (-n) + digits
    else:
        e = n - 1
        head = digits if k == 1 else f"{digits[0]}.{digits[1:]}"
        body = f"{head}e{'+' if e >= 0 else '-'}{abs(e)}"
    return sign + body


def _snake_to_camel(key: str) -> str:
    """``snake_case`` → ``camelCase``; already-camel keys pass through unchanged."""
    return re.sub(r"_([a-z0-9])", lambda m: m.group(1).upper(), key)


def _normalize_keys(value: Any) -> Any:
    """Normalize to the canonical camelCase wire rendering.

    Keys are converted snake→camel *algorithmically* (no hardcoded field
    list — future receipt fields are covered automatically) and
    ``None``-valued object keys are dropped (Python's ``None`` and TS's
    ``undefined`` both mean "field absent"). Lists keep ``None`` elements.
    """
    if isinstance(value, list):
        return [_normalize_keys(v) for v in value]
    if isinstance(value, dict):
        return {_snake_to_camel(k): _normalize_keys(v) for k, v in value.items() if v is not None}
    return value


def _canonicalize(value: Any, *, es_numbers: bool = True) -> str:
    """Deterministic JSON serialization. Mirrors the TS canonicalize().

    - Object keys sorted recursively.
    - ``None``-valued keys dropped (parity with TS dropping ``undefined``).
    - Numbers serialized ECMAScript-style via :func:`es_number` so the
      bytes match ``JSON.stringify`` cross-language (``es_numbers=False``
      reproduces the legacy ≤v0.11 ``json.dumps`` formatting for
      backward-compatible verification only).
    - Non-finite numbers (NaN, ±Inf) raise — the canonical form is total.
    """
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        if isinstance(value, float) and not math.isfinite(value):
            raise ValueError(f"canonicalize: refusing non-finite number {value}")
        return es_number(value) if es_numbers else json.dumps(value)
    if isinstance(value, str):
        return json.dumps(value, ensure_ascii=False)
    if isinstance(value, list):
        return "[" + ",".join(_canonicalize(v, es_numbers=es_numbers) for v in value) + "]"
    if isinstance(value, dict):
        keys = sorted(k for k, v in value.items() if v is not None)
        return (
            "{"
            + ",".join(
                f"{json.dumps(k, ensure_ascii=False)}:{_canonicalize(value[k], es_numbers=es_numbers)}"
                for k in keys
            )
            + "}"
        )
    raise TypeError(f"canonicalize: unsupported type {type(value).__name__}")


def canonical_receipt_payload(
    receipt: dict[str, Any],
    *,
    exclude_signed_at: bool = False,
) -> str:
    """The exact canonical byte string a receipt's signature covers.

    camelCase-normalized payload minus ``signature``/``signerPublicKey``,
    canonicalized with ES number formatting. Pass ``exclude_signed_at``
    to reproduce the legacy v0.11 form (payloads signed before
    ``signedAt`` entered the signature). Byte-for-byte identical to the
    TS ``canonicalReceiptPayload()``.
    """
    normalized = _normalize_keys(receipt)
    payload = {
        k: v
        for k, v in normalized.items()
        if k not in _SIGNATURE_FIELDS and not (exclude_signed_at and k == "signedAt")
    }
    return _canonicalize(payload)


def _legacy_snake_canonical(receipt: dict[str, Any]) -> str:
    """Legacy ≤v0.11 Python canonical form: raw snake_case keys, no
    ``signed_at``, ``json.dumps`` number formatting."""
    payload = {
        k: v
        for k, v in receipt.items()
        if k not in _SIGNATURE_FIELDS_ANY_RENDERING and v is not None
    }
    return _canonicalize(payload, es_numbers=False)


def sign_receipt(
    receipt: dict[str, Any],
    key_pair: SigningKeyPair,
    *,
    now: datetime | None = None,
) -> dict[str, Any]:
    """Return a new dict with signature, signer_public_key, signed_at populated.

    The input receipt is not mutated. The signature covers the ONE
    cross-language canonical form: the camelCase wire rendering of the
    receipt (converted algorithmically) WITH ``signedAt`` included —
    a receipt signed here verifies as-is in the TS ``verifyReceipt``
    (``ebb verify``) and vice versa. The returned dict still carries
    snake_case keys and snake_case signature fields, so Python storage
    and consumers are unchanged.
    """
    if not _HAS_CRYPTO:
        raise SigningNotInstalled()
    timestamp = (now or datetime.now(UTC)).isoformat()
    claims = {k: v for k, v in receipt.items() if k not in _SIGNATURE_FIELDS_ANY_RENDERING}
    canonical = canonical_receipt_payload({**claims, "signedAt": timestamp}).encode("utf-8")
    sig = key_pair.private_key.sign(canonical)
    return {
        **receipt,
        "signature": base64.b64encode(sig).decode("ascii"),
        "signer_public_key": key_pair.public_key_base64,
        "signed_at": timestamp,
    }


def verify_receipt(
    receipt: dict[str, Any],
    *,
    trusted_public_key: str | None = None,
) -> VerifyResult:
    """Verify a (possibly signed) receipt — mirrors the TS verifyReceipt().

    Accepts the receipt in EITHER key rendering — snake_case (Python
    ledger rows) or camelCase (TS wire shape) — keys are normalized
    algorithmically before canonicalization.

    Outcomes:
      - ``valid``           — Ed25519 signature matches the canonical payload
                              (v0.12 form with ``signedAt`` covered, or a
                              legacy v0.11 form — the reason says which).
      - ``tampered``        — signed receipt whose signature matches none of
                              the canonical forms of its fields.
      - ``key-mismatch``    — ``trusted_public_key`` set, receipt signed by someone else.
      - ``legacy-unsigned`` — pre-v0.11 receipt (or unsigned by config).
    """
    normalized = _normalize_keys(receipt)
    signature = normalized.get("signature")
    signer = normalized.get("signerPublicKey")
    if not signature or not signer:
        return VerifyResult(
            outcome="legacy-unsigned",
            reason="receipt has no `signature`/`signer_public_key` fields "
            "(pre-v0.11 or unsigned by configuration)",
        )
    if trusted_public_key and trusted_public_key != signer:
        return VerifyResult(
            outcome="key-mismatch",
            signer_public_key=signer,
            reason=f"expected signer public key {trusted_public_key}, "
            f"receipt was signed by {signer}",
        )
    if not _HAS_CRYPTO:
        raise SigningNotInstalled()
    raw_pub = base64.b64decode(signer)
    if len(raw_pub) != 32:
        raise ValueError(f"signer_public_key must be 32 bytes (got {len(raw_pub)})")
    public_key = ed25519.Ed25519PublicKey.from_public_bytes(raw_pub)
    sig_bytes = base64.b64decode(signature)

    def _matches(canonical: str) -> bool:
        try:
            public_key.verify(sig_bytes, canonical.encode("utf-8"))
        except Exception:  # InvalidSignature
            return False
        return True

    # Verification order (audit §0.3): (a) v0.12 canonical form — camelCase
    # payload WITH signedAt covered; (b) legacy v0.11 form — signedAt
    # excluded; (c) legacy ≤v0.11 Python snake_case rendering (signed the
    # snake dict as-is, json.dumps numbers). Only after every form fails →
    # tampered.
    attempts: list[tuple[str, bool]] = []

    def _push_unique(canonical: str, legacy: bool) -> None:
        if all(canonical != seen for seen, _ in attempts):
            attempts.append((canonical, legacy))

    _push_unique(canonical_receipt_payload(receipt), False)
    _push_unique(canonical_receipt_payload(receipt, exclude_signed_at=True), True)
    _push_unique(_legacy_snake_canonical(receipt), True)

    for canonical, legacy in attempts:
        if not _matches(canonical):
            continue
        reason = (
            "Ed25519 signature verified against legacy v0.11 canonical form "
            f"(signedAt not covered; {len(canonical.encode('utf-8'))} bytes)"
            if legacy
            else "Ed25519 signature verified against canonical payload "
            f"({len(canonical.encode('utf-8'))} bytes)"
        )
        return VerifyResult(outcome="valid", signer_public_key=signer, reason=reason)
    return VerifyResult(
        outcome="tampered",
        signer_public_key=signer,
        reason="Ed25519 signature does not match the canonical payload — "
        "receipt was modified after signing",
    )


__all__ = [
    "SigningKeyPair",
    "SigningNotInstalled",
    "VerifyOutcome",
    "VerifyResult",
    "canonical_receipt_payload",
    "default_signing_key_path",
    "es_number",
    "is_signing_available",
    "load_or_create_signing_key",
    "sign_receipt",
    "verify_receipt",
]

File: packages/core-ts/src/sign.ts
/**
 * Ed25519 receipt signing (v0.11+).
 *
 * Every dispatched task ships a `CarbonReceipt`. Starting in v0.11 the
 * scheduler signs each receipt with a per-installation Ed25519 keypair
 * so any consumer can verify, offline and asynchronously, that:
 *
 *   1. The receipt was produced by an `ebb-ai` installation holding
 *      the matching private key (cryptographic origin authentication).
 *   2. None of the receipt's fields have been tampered with since the
 *      moment of signing (integrity). As of v0.12 this includes
 *      `signedAt` — only `signature` and `signerPublicKey` (the proof
 *      itself) sit outside the signed payload.
 *
 * Replay defence: the signature makes `signedAt` immutable, but it does
 * NOT make a receipt unique — a holder can present the same validly
 * signed receipt twice. Consumers that care about replay must enforce
 * uniqueness ledger-side (e.g. a unique index on `taskId`, or pinning
 * each `(signerPublicKey, taskId)` pair on first sight).
 *
 * Key lifecycle is local-first by design:
 *
 *   - Keys live at `~/.ebb-ai/signing.key` (private, 0600) +
 *     `signing.key.pub` (public, 0644).
 *   - The pair is generated lazily on the first call to
 *     `loadOrCreateSigningKey()`. There is no opt-in flow — the goal
 *     is that every receipt is signable by default.
 *   - The keys are NEVER sent anywhere by ebb-ai itself. Verification
 *     consumers bundle the receipt's `signerPublicKey` along with the
 *     receipt; downstream ESG-export pipelines can pin the key on
 *     first sight.
 *
 * Cross-language canonical form (audit §0.3): there is exactly ONE
 * canonical signing payload shared by the TS and Python ports — the
 * camelCase wire rendering of the receipt, canonicalized by
 * `canonicalize()` (recursive key sort, no whitespace, ECMAScript /
 * RFC 8785-style number formatting — which `JSON.stringify` already
 * produces natively here). The Python port converts its snake_case
 * receipt dict to camelCase before canonicalizing, so a receipt signed
 * by either port verifies in both. `verifyReceipt` additionally accepts
 * a receipt in EITHER key rendering (snake_case rows read straight from
 * the shared SQLite ledger included) by normalizing keys before
 * canonicalization, and falls back to the legacy v0.11 canonical form
 * (payload without `signedAt`) for receipts signed by older releases.
 *
 * The shared fixture `test/fixtures/cross-lang-receipt-vectors.json`
 * pins the canonical bytes; both language test suites assert against
 * it byte-for-byte.
 *
 * Node ships Ed25519 in the `crypto` stdlib (since v16); there are no
 * native deps.
 */

import {
  createPrivateKey,
  createPublicKey,
  generateKeyPairSync,
  KeyObject,
  sign as nodeSign,
  verify as nodeVerify,
} from "node:crypto";
import {
  chmodSync,
  existsSync,
  mkdirSync,
  readFileSync,
  renameSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { homedir } from "node:os";
import { dirname, join } from "node:path";

import type { CarbonReceipt } from "./types.js";

/** Override the keypair location (mainly for tests). */
export interface SigningKeyOptions {
  /** Absolute path to the private key (PEM). Defaults to `~/.ebb-ai/signing.key`. */
  privateKeyPath?: string;
}

export interface SigningKeyPair {
  privateKey: KeyObject;
  publicKey: KeyObject;
  /** Base64-encoded raw 32-byte public key bytes. Suitable for embedding on a receipt. */
  publicKeyBase64: string;
  /** Filesystem path the private key lives at. */
  privateKeyPath: string;
}

/** Default location of the local signing private key. */
export function defaultSigningKeyPath(): string {
  return join(homedir(), ".ebb-ai", "signing.key");
}

/**
 * Write a file atomically: write to a same-directory temp file, then
 * `rename(2)` over the destination. Readers never observe a partial
 * file; concurrent writers converge on one winner.
 */
function writeFileAtomic(path: string, data: string, mode: number): void {
  const tmp = `${path}.tmp-${process.pid}`;
  writeFileSync(tmp, data, { mode });
  try {
    renameSync(tmp, path);
  } catch (err) {
    try { rmSync(tmp, { force: true }); } catch { /* best-effort cleanup */ }
    throw err;
  }
}

/**
 * Load the local signing keypair, generating + persisting one on first
 * call. Idempotent and safe under concurrent access (file write is
 * atomic via write-temp-then-rename — single writer wins, both readers
 * converge).
 */
export function loadOrCreateSigningKey(
  opts: SigningKeyOptions = {},
): SigningKeyPair {
  const privateKeyPath = opts.privateKeyPath ?? defaultSigningKeyPath();
  const publicKeyPath = `${privateKeyPath}.pub`;

  if (existsSync(privateKeyPath)) {
    const privatePem = readFileSync(privateKeyPath, "utf8");
    const privateKey = createPrivateKey({ key: privatePem, format: "pem" });
    const publicKey = createPublicKey(privateKey);
    return {
      privateKey,
      publicKey,
      publicKeyBase64: rawPublicKeyBase64(publicKey),
      privateKeyPath,
    };
  }

  const { privateKey, publicKey } = generateKeyPairSync("ed25519");
  const privatePem = privateKey.export({ type: "pkcs8", format: "pem" }) as string;
  const publicPem = publicKey.export({ type: "spki", format: "pem" }) as string;

  mkdirSync(dirname(privateKeyPath), { recursive: true });
  writeFileAtomic(privateKeyPath, privatePem, 0o600);
  writeFileAtomic(publicKeyPath, publicPem, 0o644);
  // Defensive re-chmod in case umask interfered.
  try { chmodSync(privateKeyPath, 0o600); } catch { /* best-effort */ }

  return {
    privateKey,
    publicKey,
    publicKeyBase64: rawPublicKeyBase64(publicKey),
    privateKeyPath,
  };
}

/**
 * Deterministic JSON canonicalization. Recursively sorts object keys;
 * stringifies numbers via `JSON.stringify` (which already produces the
 * shortest round-trip, RFC 8785 / JCS-compatible representation for
 * finite floats); rejects non-finite numbers and `undefined` so the
 * canonical form is total.
 */
export function canonicalize(value: unknown): string {
  if (value === null) return "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") {
    if (!Number.isFinite(value)) {
      throw new Error(`canonicalize: refusing to serialize non-finite number ${value}`);
    }
    return JSON.stringify(value);
  }
  if (typeof value === "string") return JSON.stringify(value);
  if (Array.isArray(value)) {
    return `[${value.map(canonicalize).join(",")}]`;
  }
  if (typeof value === "object") {
    const obj = value as Record<string, unknown>;
    const keys = Object.keys(obj).filter((k) => obj[k] !== undefined).sort();
    const parts = keys.map((k) => `${JSON.stringify(k)}:${canonicalize(obj[k])}`);
    return `{${parts.join(",")}}`;
  }
  throw new Error(`canonicalize: unsupported value of type ${typeof value}`);
}

/**
 * Fields excluded from the signed payload — they are the proof itself.
 * Note `signedAt` is NOT here: since v0.12 it is covered by the
 * signature (audit §0.8 — a free-floating `signedAt` was trivially
 * rewritable despite being documented as replay defence).
 */
const SIGNATURE_FIELDS = new Set(["signature", "signerPublicKey"]);

/** Every signature-machinery key, in both wire renderings. */
const SIGNATURE_FIELDS_ANY_RENDERING = new Set([
  "signature",
  "signerPublicKey",
  "signer_public_key",
  "signedAt",
  "signed_at",
]);

/** `snake_case` → `camelCase`; already-camel keys pass through unchanged. */
function snakeToCamelKey(key: string): string {
  return key.replace(/_([a-z0-9])/g, (_m, c: string) => c.toUpperCase());
}

/**
 * Normalize a receipt-shaped value to the canonical camelCase wire
 * rendering: keys are converted snake→camel algorithmically (so future
 * receipt fields are covered automatically) and `null` / `undefined`
 * valued object keys are dropped (Python's `None` and TS's `undefined`
 * both mean "field absent" on a receipt). Arrays keep `null` elements.
 */
function normalizeReceiptKeys(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(normalizeReceiptKeys);
  if (value !== null && typeof value === "object") {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      if (v === undefined || v === null) continue;
      out[snakeToCamelKey(k)] = normalizeReceiptKeys(v);
    }
    return out;
  }
  return value;
}

/**
 * The exact canonical byte string a receipt's signature covers:
 * camelCase-normalized payload minus `signature`/`signerPublicKey`,
 * canonicalized. Pass `excludeSignedAt` to reproduce the legacy v0.11
 * form (payloads signed before `signedAt` entered the signature).
 *
 * Exported so tests (and external verifiers) can pin canonical bytes;
 * shared cross-language test vectors assert this byte-for-byte against
 * the Python port.
 */
export function canonicalReceiptPayload(
  receipt: Record<string, unknown>,
  opts: { excludeSignedAt?: boolean } = {},
): string {
  const normalized = normalizeReceiptKeys(receipt) as Record<string, unknown>;
  const payload: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(normalized)) {
    if (SIGNATURE_FIELDS.has(k)) continue;
    if (opts.excludeSignedAt && k === "signedAt") continue;
    payload[k] = v;
  }
  return canonicalize(payload);
}

/**
 * Legacy v0.11 canonical form of the receipt's ORIGINAL key rendering
 * (no snake→camel normalization, no `signedAt`). Python ≤v0.11 signed
 * its snake_case dict directly; this lets those ledger rows keep
 * verifying.
 */
function legacyRawCanonical(receipt: Record<string, unknown>): string {
  const payload: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(receipt)) {
    if (SIGNATURE_FIELDS_ANY_RENDERING.has(k)) continue;
    if (v === undefined || v === null) continue;
    payload[k] = v;
  }
  return canonicalize(payload);
}

/**
 * Sign a receipt in-place-style: returns a new receipt with the three
 * signature fields populated. The original receipt is not mutated.
 *
 * `signedAt` is computed first and INCLUDED in the signed payload —
 * only `signature` and `signerPublicKey` are excluded.
 */
export function signReceipt(
  receipt: CarbonReceipt,
  keyPair: SigningKeyPair,
  now: () => Date = () => new Date(),
): CarbonReceipt {
  const signedAt = now().toISOString();
  const claims: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(receipt)) {
    if (SIGNATURE_FIELDS_ANY_RENDERING.has(k)) continue;
    claims[k] = v;
  }
  const canonical = canonicalReceiptPayload({ ...claims, signedAt });
  const sig = nodeSign(null, Buffer.from(canonical, "utf8"), keyPair.privateKey);
  return {
    ...receipt,
    signature: sig.toString("base64"),
    signerPublicKey: keyPair.publicKeyBase64,
    signedAt,
  };
}

export type VerifyOutcome = "valid" | "tampered" | "legacy-unsigned" | "key-mismatch";

export interface VerifyResult {
  outcome: VerifyOutcome;
  /** Public key bytes that signed the receipt (base64). undefined for legacy-unsigned. */
  signerPublicKey?: string;
  /** Reason text suitable for CLI / API surfacing. */
  reason: string;
}

/**
 * Verify a (possibly signed) receipt. Accepts the receipt in EITHER key
 * rendering — camelCase (TS wire shape) or snake_case (Python ledger
 * rows) — keys are normalized algorithmically before canonicalization.
 *
 *   - `valid`           — signature matches the canonical payload
 *                         (v0.12 form with `signedAt` covered, or the
 *                         legacy v0.11 form — the reason says which).
 *   - `tampered`        — receipt is signed but the signature matches
 *                         none of the canonical forms of its fields.
 *   - `key-mismatch`    — caller supplied a `trustedPublicKey` that
 *                         differs from the one bundled on the receipt.
 *   - `legacy-unsigned` — pre-v0.11 receipt (no signature). The
 *                         consumer must decide whether to accept it.
 */
export function verifyReceipt(
  receipt: CarbonReceipt | Record<string, unknown>,
  opts: { trustedPublicKey?: string } = {},
): VerifyResult {
  const raw = receipt as Record<string, unknown>;
  const normalized = normalizeReceiptKeys(raw) as Record<string, unknown>;
  const signature = normalized.signature as string | undefined;
  const signerPublicKey = normalized.signerPublicKey as string | undefined;
  if (!signature || !signerPublicKey) {
    return {
      outcome: "legacy-unsigned",
      reason: "receipt has no `signature` / `signerPublicKey` fields (pre-v0.11 or unsigned by configuration)",
    };
  }
  if (opts.trustedPublicKey && opts.trustedPublicKey !== signerPublicKey) {
    return {
      outcome: "key-mismatch",
      signerPublicKey,
      reason: `expected signer public key ${opts.trustedPublicKey}, receipt was signed by ${signerPublicKey}`,
    };
  }
  const publicKey = publicKeyFromBase64Raw(signerPublicKey);
  const sigBytes = Buffer.from(signature, "base64");
  const matches = (canonical: string): boolean =>
    nodeVerify(null, Buffer.from(canonical, "utf8"), publicKey, sigBytes);

  // Verification order (audit §0.3): (a) v0.12 canonical form — camelCase
  // payload WITH signedAt covered; (b) legacy v0.11 form — signedAt
  // excluded; (c) legacy v0.11 snake_case rendering (Python ≤v0.11 signed
  // its snake dict as-is). Only after every form fails → tampered.
  const attempts: Array<{ canonical: string; legacy: boolean }> = [];
  const pushUnique = (canonical: string, legacy: boolean): void => {
    if (!attempts.some((a) => a.canonical === canonical)) attempts.push({ canonical, legacy });
  };
  pushUnique(canonicalReceiptPayload(raw), false);
  pushUnique(canonicalReceiptPayload(raw, { excludeSignedAt: true }), true);
  pushUnique(legacyRawCanonical(raw), true);

  for (const attempt of attempts) {
    if (!matches(attempt.canonical)) continue;
    return {
      outcome: "valid",
      signerPublicKey,
      reason: attempt.legacy
        ? `Ed25519 signature verified against legacy v0.11 canonical form (signedAt not covered; ${attempt.canonical.length} bytes)`
        : `Ed25519 signature verified against canonical payload (${attempt.canonical.length} bytes)`,
    };
  }
  return {
    outcome: "tampered",
    signerPublicKey,
    reason: "Ed25519 signature does not match the canonical payload — receipt was modified after signing",
  };
}

// --------------------------------------------------------------------------- //
// Helpers

function rawPublicKeyBase64(publicKey: KeyObject): string {
  // Node 16+ exposes a 'raw' jwk-ish format via the JWK export.
  const jwk = publicKey.export({ format: "jwk" }) as { x?: string };
  if (!jwk.x) {
    throw new Error("loadOrCreateSigningKey: failed to extract raw Ed25519 public key bytes");
  }
  // JWK encodes `x` as base64url; normalize to base64 so the receipt is
  // self-consistent with `Buffer.from(sig, "base64")` on the verify side.
  return Buffer.from(jwk.x, "base64url").toString("base64");
}

function publicKeyFromBase64Raw(b64: string): KeyObject {
  const raw = Buffer.from(b64, "base64");
  if (raw.length !== 32) {
    throw new Error(`verifyReceipt: signerPublicKey must be 32 bytes (got ${raw.length})`);
  }
  return createPublicKey({
    key: { kty: "OKP", crv: "Ed25519", x: raw.toString("base64url") },
    format: "jwk",
  });
}
